support removal of event handlers from SharedIndexInformers

alexzielenski opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Text from original PR:

```
To be able to implement controllers that are dynamically deciding
on which resources to watch, it is required to get rid of
dedicated watches and event handlers again. This requires the
possibility to remove event handlers from SharedIndexInformers again.
Stopping an informer is not sufficient, because there might
be multiple controllers in a controller manager that independently
decide which resources to watch.

Unfortunately the ResourceEventHandler interface encourages to use
value objects for handlers (like the ResourceEventHandlerFuncs
struct, that uses value receivers to implement the interface).
Go does not support comparison of function pointers and therefore
the comparison of such structs is not possible, also. To be able
to remove all kinds of handlers and to solve the problem of
multi-registrations of handlers a registration handle is introduced.
It is returned when adding a handler and can later be used to remove
the registration again. This handle directly stores the created
listener to simplify the deletion.
```

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

Continuation of #98657 after work there no longer took place.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/cc @FillZpp 
/cc @kevindelgado

mandelsoft committed:
rename handle to registration

mandelsoft committed:
apply desired changes for handler registration

mandelsoft committed:
tests for invalid registration removals

mandelsoft committed:
active remove/add tests for event handlers

alexzielenski committed:
support removal of event handlers from SharedIndexInformers

To be able to implement controllers that are dynamically deciding
on which resources to watch, it is required to get rid of
dedicated watches and event handlers again. This requires the
possibility to remove event handlers from SharedIndexInformers again.
Stopping an informer is not sufficient, because there might
be multiple controllers in a controller manager that independently
decide which resources to watch.

Unfortunately the ResourceEventHandler interface encourages to use
value objects for handlers (like the ResourceEventHandlerFuncs
struct, that uses value receivers to implement the interface).
Go does not support comparison of function pointers and therefore
the comparison of such structs is not possible, also. To be able
to remove all kinds of handlers and to solve the problem of
multi-registrations of handlers a registration handle is introduced.
It is returned when adding a handler and can later be used to remove
the registration again. This handle directly stores the created
listener to simplify the deletion.

alexzielenski committed:
remove informational informer methods again

alexzielenski committed:
switch listeners to use a map, adapt tests

alexzielenski commented:
/sig api-machinery
/triage accepted

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -619,38 +656,77 @@ func (s *sharedIndexInformer) OnDelete(old interface{}) {
+	wg        wait.Group
+}
+
+func (p *sharedProcessor) getListener(registration ResourceEventHandlerRegistration) *processorListener {
```
This is only used by tests right? I would maybe comment that just so it's clear

kevindelgado reviewed (commented):
This looks pretty good to me!

I was struggling to give meaningful feedback, mostly because I was trying to understand the diff between this and where mandelsoft left off but it’s a little hard because a lot has changed between then and now.

My understanding is that the main changes are:
1. Using the opaque interface as the registration handle
2. Turning the shared processor’s listeners slice into a map of listener to a bool of if they’re syncing (making the syncing listeners slice redundant).

Let me know if I’m missing anything else, but overall looks good, I’ll leave it for others to review.

alexzielenski commented:
/retest

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -660,19 +736,24 @@ func (p *sharedProcessor) run(stopCh <-chan struct{}) {
-	defer p.listenersLock.RUnlock()
-	for _, listener := range p.listeners {
+
+	p.listenersLock.Lock()
```
How long does that Wait() take on line 752? Does this lock get held the entire time while the informer is in operation?

alexzielenski replied:
I've removed this as a possible issue by placing the locking in an anonymous function scope, so the wait happens without holding any lock

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -196,59 +215,60 @@ func TestResyncCheckPeriod(t *testing.T) {
 		t.Errorf("expected %d, got %d", e, a)
 	}
-	if e, a := time.Duration(0), informer.processor.listeners[0].resyncPeriod; e != a {
+	if e, a := time.Duration(0), informer.processor.getListener(listener1Registration).resyncPeriod; e != a {
```
optional nit: for readability I'd consider declaring `gl := informer.processor.getListener` so that the `informer.processor.getListener(listener1Registration)` lines can be much shorter

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -196,59 +215,60 @@ func TestResyncCheckPeriod(t *testing.T) {
 	// listener 2, resync every minute
 	listener2 := newTestListener("listener2", 1*time.Minute)
-	informer.AddEventHandlerWithResyncPeriod(listener2, listener2.resyncPeriod)
+	listener2Registration, _ := informer.AddEventHandlerWithResyncPeriod(listener2, listener2.resyncPeriod)
```
also extremely optional nit, personally I'd probably have called these vars `h2` etc

lavalamp reviewed (commented):
I'm a little concerned about the locking -- often for stuff like this I like to have a test that repeatedly and randomly adds and removes thingies to look for locking errors. Given the tests you added, it's probably not mandatory if you're super confident, but I think that there's a reasonable chance adding this could e.g. expose an existing such bug?

alexzielenski committed:
address review comments

asd

alexzielenski committed:
remove duplicate test

became a duplicate while refactoring original PR

alexzielenski committed:
add multithreaded test to shared informer

alexzielenski committed:
reset listenersStarted

for correctness. technically shouldnt be an issue since restarting a stopped processor is not supported

apelisse commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -619,38 +656,77 @@ func (s *sharedIndexInformer) OnDelete(old interface{}) {
-	} else {
-		for _, listener := range p.listeners {
+	for listener, isSyncing := range p.listeners {
+		if !sync || isSyncing {
```
This is not doing the same thing as before. i.e., this is now also adding non-syncing objects when sync is false. I have no idea if this is important.

alexzielenski replied:
That is intentional and I believe it is equivalent to the original semantics. The original code would add the object to all syncing listeners if `sync` was true. If `sync` was false the original code added the object to all listeners including syncing listeners.

In the new code if the object is added to only syncing listeners if `sync` is `true`. If `sync` is `false` it is added to all listeners as before.

apelisse commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -660,18 +736,30 @@ func (p *sharedProcessor) run(stopCh <-chan struct{}) {
-	}
+
+	func() {
+		p.listenersLock.Lock()
```
I'm assuming you moved that to its own function to unlock before the final wg wait?

alexzielenski replied:
Yes, however it is important that the wg is not used after releasing the lock. Since the `wg` is only used when `listenersStarted` is true, it is not a problem. However, I think it is a bit sensitive.

apelisse commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -681,28 +769,30 @@ func (p *sharedProcessor) shouldResync() bool {
+
+		if shouldResync {
 			resyncNeeded = true
-			p.syncingListeners = append(p.syncingListeners, listener)
```
This used to copy listener into syncing (rather than move). Maybe that's why the code above is fine?

alexzielenski replied:
Yes, the syncingListeners list is the subset of `listeners` which should receive sync events.

apelisse reviewed (commented):
Thanks, I've looked at the code, will take a look at the tests next.

apelisse commented:
Looks good to me, I'll let someone more knowledgable of this area tag for me. Thanks!

alexzielenski committed:
fix spelling

alexzielenski commented:
/assign @deads2k

deads2k commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -619,38 +656,77 @@ func (s *sharedIndexInformer) OnDelete(old interface{}) {
-	} else {
-		for _, listener := range p.listeners {
+	for listener, isSyncing := range p.listeners {
+		if !sync || isSyncing {
```
this logic looks to fullfill the same intent, but it was hard for me to read.  How about

```go
switch{
case !sync:
  // non-sync messages are delivered to every listener
  listener.add(obj)
case isSyncing:
  // sync messages are delivered to every syncing listenter
  listener.add(obj)
default:
  // skipping a sync obj for a non-syncing listener
}
```

deads2k commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -660,18 +736,30 @@ func (p *sharedProcessor) run(stopCh <-chan struct{}) {
-		close(listener.addCh) // Tell .pop() to stop. .pop() will tell .run() to stop
-	}
+
+	func() {
```
What was the problem with holding lock through the wg.Wait?  Holding the lock until the shutdown was complete prevented any other thread from adding listeners while shutting down.  Without that protection, what stops some from adding listeners during the shutdown sequence?

alexzielenski replied:
It was a mistake on my end. I had intended to avoid holding the lock for the entire duration while the sharedInformer ran as per https://github.com/kubernetes/kubernetes/pull/111122#discussion_r930320473. Revisiting this now I realize the lock is only held while the informer is actively stopping. 

I've changed this bit so that the lock is held until the informers all stop.

alexzielenski committed:
simplify control flow

alexzielenski committed:
hold listener lock while waiting for goroutines to finish

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -629,6 +632,121 @@ func TestRemovingRemovedSharedInformer(t *testing.T) {
+					err := informer.RemoveEventHandler(handles[idx])
+					if err != nil {
+						if strings.Contains(err.Error(), "stopped already") {
+							continue
```
shouldn't you still execute the removal below in this case?

alexzielenski replied:
If the error was "stopped already" then the test is over and <-ctx.Done() case would stop the worker the next iteration.  I've changed the `continue` to just `return` directly here to make that more clear

lavalamp commented:
only one minor comment, probably ready to go otherwise?

alexzielenski committed:
return when test is done

lavalamp commented:
/approve

alexzielenski commented:
/retest

apelisse commented:
/lgtm

Thanks!
